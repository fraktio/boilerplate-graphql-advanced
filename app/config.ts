/* eslint-disable no-process-env */

export const API_PORT = "API_PORT";
export const STDOUT_LOGGING = "STDOUT_LOGGING";
export const PRODUCTION = "PRODUCTION";
export const API_CORS_ENDPOINT = "API_CORS_ENDPOINT";
export const TOKEN_PATH = "TOKEN_PATH";
export const TOKEN_DOMAIN = "TOKEN_DOMAIN";
export const TOKEN_SECRET = "TOKEN_SECRET";
export const ACCESS_TOKEN_AGE_SECONDS = "ACCESS_TOKEN_AGE_SECONDS";
export const REFRESH_TOKEN_AGE_SECONDS = "REFRESH_TOKEN_AGE_SECONDS";
export const DATABASE_TYPE = "DATABASE_TYPE";
export const DATABASE_HOST = "DATABASE_HOST";
export const DATABASE_USER = "DATABASE_USER";
export const DATABASE_PORT = "DATABASE_PORT";
export const DATABASE_PASSWORD = "DATABASE_PASSWORD";
export const DATABASE_DATABASE_NAME = "DATABASE_DATABASE_NAME";
export const NUMBER_API_MOCK_TOKEN = "NUMBER_API_MOCK_TOKEN";

export const getEnvIntFallback = (envKey: string, fallback: number): number => {
  const value = process.env[envKey];

  if (!value) {
    return fallback;
  }

  const parsed = parseInt(value, 10);

  if (isNaN(parsed)) {
    throw new Error(
      `Environment variable ${envKey}. Received invalid value for INT: ${value}`,
    );
  }

  return parsed;
};

export const getEnvFallback = (envKey: string, fallback: string): string => {
  const value = process.env[envKey];

  if (!value) {
    return fallback;
  }

  return value;
};

const requireEnv = (env: string): string => {
  const envVariable = process.env[env];

  if (!envVariable) {
    throw new Error(`Environment variable ${env} is missing!`);
  }

  return envVariable;
};

const requireIntEnv = (env: string): number => {
  const value = requireEnv(env);

  const parsed = parseInt(value, 10);

  if (!parsed || isNaN(parsed)) {
    throw new Error(`Environment variable ${env} is missing!`);
  }

  return parsed;
};

const requireBoolean = (env: string): boolean => {
  const value = requireEnv(env);

  if (value === "TRUE") {
    return true;
  }

  if (value === "FALSE") {
    return false;
  }

  throw new Error(
    `Environment variable ${env} only accepts 'TRUE' or 'FALSE'!`,
  );
};

export const getCookiesConfigFromEnv = () => ({
  path: requireEnv(TOKEN_PATH),
  domain: requireEnv(TOKEN_DOMAIN),
  secret: requireEnv(TOKEN_SECRET),
  accessAgeSeconds: requireIntEnv(ACCESS_TOKEN_AGE_SECONDS),
  refreshAgeSeconds: requireIntEnv(REFRESH_TOKEN_AGE_SECONDS),
});

export type CookiesConfig = ReturnType<typeof getCookiesConfigFromEnv>;

export const getDatabaseConfigFromEnv = () => ({
  type: requireEnv(DATABASE_TYPE),
  host: requireEnv(DATABASE_HOST),
  user: requireEnv(DATABASE_USER),
  port: requireIntEnv(DATABASE_PORT),
  password: requireEnv(DATABASE_PASSWORD),
  databaseName: requireEnv(DATABASE_DATABASE_NAME),
});

export type DatabaseConfig = ReturnType<typeof getDatabaseConfigFromEnv>;

export const getNumberFactConfigFromEnv = () => ({
  token: requireEnv(NUMBER_API_MOCK_TOKEN),
});

export type NumberFactConfig = ReturnType<typeof getNumberFactConfigFromEnv>;

export const getConfigFromEnv = () => ({
  apiPort: requireIntEnv(API_PORT),
  stdoutLogging: requireBoolean(STDOUT_LOGGING),
  isProduction: requireBoolean(PRODUCTION),
  apiCorsEndpoint: requireEnv(API_CORS_ENDPOINT),

  cookies: getCookiesConfigFromEnv(),
  database: getDatabaseConfigFromEnv(),

  numberFact: getNumberFactConfigFromEnv(),
});

export type Config = ReturnType<typeof getConfigFromEnv>;
