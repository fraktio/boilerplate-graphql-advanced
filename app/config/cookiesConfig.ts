import * as t from "io-ts";

import { NumberFromString } from "~/utils/decoders";

export const CookiesConfigDecoder = t.type({
  path: t.string,
  domain: t.string,
  secret: t.string,
  accessTokenAgeSeconds: NumberFromString,
});

export type CookiesConfig = t.TypeOf<typeof CookiesConfigDecoder>;
