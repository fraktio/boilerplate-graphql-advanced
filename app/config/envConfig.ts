import * as t from "io-ts";

import { BooleanFromString, NumberFromString } from "~/utils/decoders";

export const EnvConfigDecoder = t.type({
  apiPort: NumberFromString,
  isProduction: BooleanFromString,
  apiCorsEndpoint: t.string,
});

export type EnvConfig = t.TypeOf<typeof EnvConfigDecoder>;
