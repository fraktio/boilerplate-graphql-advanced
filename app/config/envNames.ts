export const LOGGING_LEVEL = "LOGGING_LEVEL";
export const API_PORT = "API_PORT";
export const IS_PRODUCTION = "IS_PRODUCTION";
export const API_CORS_ENDPOINT = "API_CORS_ENDPOINT";
export const API_EXPOSE_ERRORS = "API_EXPOSE_ERRORS";

export const TOKEN_PATH = "TOKEN_PATH";
export const TOKEN_DOMAIN = "TOKEN_DOMAIN";
export const TOKEN_SECRET = "TOKEN_SECRET";
export const ACCESS_TOKEN_AGE_SECONDS = "ACCESS_TOKEN_AGE_SECONDS";
export const REFRESH_TOKEN_AGE_SECONDS = "REFRESH_TOKEN_AGE_SECONDS";

// Graphql
export const GRAPHQL_ALLOW_INTROSPECTION = "GRAPHQL_ALLOW_INTROSPECTION";
export const GRAPHQL_INLINE_TRACE = "GRAPHQL_INLINE_TRACE";

// APIs
export const API_NUMBER_FACT_TOKEN = "API_NUMBER_FACT_TOKEN";

// Database config
export const DATABASE_TYPE = "DATABASE_TYPE";
export const DATABASE_HOST = "DATABASE_HOST";
export const DATABASE_USER = "DATABASE_USER";
export const DATABASE_PORT = "DATABASE_PORT";
export const DATABASE_PASSWORD = "DATABASE_PASSWORD";
export const DATABASE_NAME = "DATABASE_NAME";

// Redis
export const IS_REDIS_USED = "IS_REDIS_USED";
export const REDIS_HOST = "REDIS_HOST";
export const REDIS_PORT = "REDIS_PORT";
export const REDIS_USERNAME = "REDIS_USERNAME";
export const REDIS_PASSWORD = "REDIS_PASSWORD";

// Platform config
export const PLATFORM = "PLATFORM";
export const GOOGLE_CLOUD_SQL_CONNECTION_NAME =
  "GOOGLE_CLOUD_SQL_CONNECTION_NAME";
export const GOOGLE_CLOUD_PROJECT_ID = "GOOGLE_CLOUD_PROJECT_ID";
