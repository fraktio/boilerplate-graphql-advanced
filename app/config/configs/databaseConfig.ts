import {
  DATABASE_NAME,
  DATABASE_HOST,
  DATABASE_PASSWORD,
  DATABASE_PORT,
  DATABASE_TYPE,
  DATABASE_USER,
} from "~/config/envNames";
import { getEnv, getEnvInt } from "~/config/getters";

const allowedDatabaseTypes = ["pg" as const];

type DatabaseType = typeof allowedDatabaseTypes[number];

export type DatabaseConfig = {
  type: DatabaseType;
  host: string;
  user: string;
  port: number;
  password: string;
  databaseName: string;
};

export const createDatabaseConfig = (): DatabaseConfig => ({
  type: getEnv(DATABASE_TYPE, allowedDatabaseTypes),
  host: getEnv(DATABASE_HOST),
  user: getEnv(DATABASE_USER),
  port: getEnvInt(DATABASE_PORT),
  password: getEnv(DATABASE_PASSWORD),
  databaseName: getEnv(DATABASE_NAME),
});
