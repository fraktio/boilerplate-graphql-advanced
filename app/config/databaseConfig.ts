import { isLeft } from "fp-ts/lib/Either";
import * as t from "io-ts";
import reporter from "io-ts-reporters";

import { getEnv } from "~/config/config";
import {
  DATABASE_DATABASE_NAME,
  DATABASE_HOST,
  DATABASE_PASSWORD,
  DATABASE_PORT,
  DATABASE_TYPE,
  DATABASE_USER,
} from "~/config/variablesConfig";
import { NumberFromString } from "~/utils/decoders";

export const DatabaseConfigDecoder = t.type({
  type: t.string,
  host: t.string,
  user: t.string,
  port: NumberFromString,
  password: t.string,
  databaseName: t.string,
});

export type DatabaseConfig = t.TypeOf<typeof DatabaseConfigDecoder>;

export const validateDatabaseConfig = (config: unknown): DatabaseConfig => {
  const validated = DatabaseConfigDecoder.decode(config);

  if (isLeft(validated)) {
    const errors = reporter.report(validated);
    console.error(errors.join("\n"));
    throw new Error("Invalid environmental configation");
  }

  return validated.right;
};

export const createDatabaseConfig = (): DatabaseConfig => {
  const dbConfig = {
    type: getEnv(DATABASE_TYPE),
    host: getEnv(DATABASE_HOST),
    user: getEnv(DATABASE_USER),
    port: getEnv(DATABASE_PORT),
    password: getEnv(DATABASE_PASSWORD),
    databaseName: getEnv(DATABASE_DATABASE_NAME),
  };

  return validateDatabaseConfig(dbConfig);
};
