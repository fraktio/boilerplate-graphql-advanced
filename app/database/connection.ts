import knex from "knex";
import type Knex from "knex";

import { Config } from "~/config";

export type DBConnection = Knex;

export const createKnex = (opts: { config: Config }) =>
  knex(getConnection({ config: opts.config }));

export const testConnection = async (opts: { knex: Knex }) => {
  await opts.knex.raw("SELECT 1 + 1 as result;");
};

export const getConnection = (opts: { config: Config }) => ({
  client: opts.config.databaseType,
  connection: {
    host: opts.config.databaseHost,
    user: opts.config.databaseUser,
    port: opts.config.databasePort,
    password: opts.config.databasePassword,
    database: opts.config.databaseDatabaseName,
  },
  migrations: {
    directory: "./migrations",
  },
  seeds: {
    directory: "./seeds",
  },
  useNullAsDefault: true,
});
