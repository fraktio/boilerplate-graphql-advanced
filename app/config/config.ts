import { APIConfig, createAPIConfig } from "~/config/configs/apiConfig";
import {
  createDatabaseConfig,
  DatabaseConfig,
} from "~/config/configs/databaseConfig";
import { createEnvConfig, EnvConfig } from "~/config/configs/envConfig";
import {
  createLoggingConfig,
  LoggingConfig,
} from "~/config/configs/loggingConfig";
import {
  createPlatformConfig,
  PlatformConfig,
} from "~/config/configs/platformConfig";
import {
  createSessionConfig,
  SessionConfig,
} from "~/config/configs/sessionConfig";

export type Config = {
  api: APIConfig;
  database: DatabaseConfig;
  env: EnvConfig;
  logging: LoggingConfig;
  platform: PlatformConfig;
  session: SessionConfig;
};

export const createConfig = (): Config => ({
  api: createAPIConfig(),
  database: createDatabaseConfig(),
  env: createEnvConfig(),
  logging: createLoggingConfig(),
  platform: createPlatformConfig(),
  session: createSessionConfig(),
});
