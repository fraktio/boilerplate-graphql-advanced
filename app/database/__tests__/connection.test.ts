import { getConnection } from "../connection";

import { DatabaseConfig } from "~/config/databaseConfig";

describe("database tests / connection", () => {
  it("createUUID", async () => {
    const databaseConfig: DatabaseConfig = {
      type: "type",
      host: "host",
      user: "user",
      port: 5432,
      password: "password",
      databaseName: "databaseName",
    };

    const result = getConnection({ databaseConfig });

    expect(result).toMatchSnapshot();
  });
});
