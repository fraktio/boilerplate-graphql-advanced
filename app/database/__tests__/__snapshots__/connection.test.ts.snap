// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`database tests / connection createUUID 1`] = `
Object {
  "client": "pg",
  "connection": Object {
    "database": "databaseName",
    "host": "host",
    "password": "password",
    "port": 5432,
    "user": "user",
  },
  "migrations": Object {
    "directory": "./migrations",
  },
  "seeds": Object {
    "directory": "./seeds",
  },
  "useNullAsDefault": true,
}
`;
